Docs: revamp Intro/Install (#4060)
aiidateam/aiida-core: docs/source/intro/get_started.rst

.. _intro:get_started:

****************
Getting Started
****************

A working AiiDA installation consists of three core components, plus any external codes you wish to run:

> * ``aiida-core``: The main python package and associated CLI ``verdi``.

Reviewer: ```suggestion
* aiida-core: The main Python package and the associated command line tool: ``verdi``.
```

> * |PostgreSQL|: A service which manages the database where we store generated data.

Reviewer: ```suggestion
* |PostgreSQL|: A service which manages the database that AiiDA uses to store generated data.
```

> * |RabbitMQ|: A service which manages communication with the processes that we run.

Reviewer: ```suggestion
* |RabbitMQ|: A service which manages communication with the processes that AiiDA spawns.
```

> AiiDA provides a number of routes to setting up this environment, depending on your use case.
> These are outlined below, followed by a recommended "quick-install" route on your local computer.

Reviewer: ```suggestion
There are multiple routes to setting up a working AiiDA environment.
These are shown below, followed by a recommended "quick-install" route on your local computer.
```

.. panels::
   :body: bg-light
   :footer: bg-light border-0

   :fa:`desktop,mr-1` **Direct (Bare Metal)**

>    *Install software directly into your local root directory.*

Reviewer: Does "root directory" here imply that the "Direct" installation method is different from installing *via* PyPI or conda? What is a pip/conda installation with only local directories then?

Especially using conda, it is possible (and should be encouraged) to install the system services in user space. I think it should be very clear that the first panel is essentially equivalent to following the rest of the instructions on this page.

Author: The key difference IMO between the first and second panel is that, with a direct install you can only install a **singular** instance of the software (well technically a single instance per user, but let's not complicate matters further!).

Going the conda route, means that, yes you install conda itself "directly", but then **everything** else is installed in a virtual environment (including postgres and rabbitmq). This IMO is not equivalent to the first panel, it is primarily the second panel.

With only pip; you have to install postgres and rabbitmq directly. Then you could either use pip directly to install python packages, which would be a "full" direct installation (discouraged), or you can use venv/pipenv to install the python packages in an isolated folder, which is kind of 50/50 between panel one and two.

I have re-worded slightly in 9c0599e to maybe make this clearer

Reviewer: Since using conda or pip in a venv is what we recommend, should we make that the first panel then? A direct "bare metal" installation into should be either the second or potentially even the last option IMO.

Reviewer: Also, I think we should try to somehow visually highlight our recommended installation method.

Author: > A direct "bare metal" installation into should be either the second or potentially even the last option IMO.

The way I laid it out was for "increasing virtualization" as opposed to "most recommended", IMO virtual environment doesn't really make sense first, since it still involves an element of bare metal installation. 

> I think we should try to somehow visually highlight our recommended installation method.

Is that not highlighted by the subsequent Quick Start section?

Reviewer: > > A direct "bare metal" installation into should be either the second or potentially even the last option IMO.
> 
> The way I laid it out was for "increasing virtualization" as opposed to "most recommended", IMO virtual environment doesn't really make sense first, since it still involves an element of bare metal installation.

Fair enough.

> > I think we should try to somehow visually highlight our recommended installation method.
> 
> Is that not highlighted by the subsequent Quick Start section?

I think it's important to have a *recommended* way that is clearly marked. But if we can't agree on that, I'm also ok to just leave it *as-is* for now.

Author: Yeh I'm going to leave as-is for now. Plus I would note that the line above the panels says: These are shown below, followed by a recommended “quick-install” route on your local computer.

   :link-badge:`https://anaconda.org/conda-forge/aiida-core,aiida-core on Conda,url,badge-primary text-white`

   ---------------

   :fa:`cube,mr-1` **Containers**

   *Use a pre-made image of all the required software.*

   AiiDA maintains a `Docker <https://www.docker.com/>`__ image, which is particularly useful for learning and testing purposes.
   It is a great way to quickly get started on the tutorials.

   +++

   :link-badge:`intro:install:docker,Docker Tutorial,ref,badge-primary text-white`
   :link-badge:`https://hub.docker.com/r/aiidateam/aiida-core,aiida-core on DockerHub,url,badge-primary text-white`

   ---------------

   :fa:`cloud,mr-1` **Virtual Machines**

   *Use a pre-made machine with all the required software.*

   `Materials Cloud <https://www.materialscloud.org>`__ provides both downloadable and web based VMs,
   also incorporating pre-installed Materials Science codes.

   +++

   :link-badge:`https://materialscloud.org/quantum-mobile,Quantum Mobile,url,badge-primary text-white`
   :link-badge:`https://aiidalab.materialscloud.org,AiiDA lab,url,badge-primary text-white`

.. _intro:quick_start:

Quick Start Install

> ===================

Reviewer: I think it would be good to divide this guide into multiple steps, something like:

1. Install aiida-core
2. Initialize/setup(?) database
3. Setup profile
4. Start services
5. Stop services

Author: Fixed in e5c13be

    ----------------------------------------------

    :fa:`download,mr-1` **Install with pip**

    .. code-block:: console

        $ pip install aiida-core
        $ reentry scan

    ``aiida-core`` can be installed from `PyPi <https://pypi.org/project/aiida-core>`__.
    You will then need to install |PostgreSQL| and |RabbitMQ| depending on your operating system.

    :link-badge:`intro:install:prerequisites,Install prerequisites,ref,badge-primary text-white`

Before working with AiiDA, you must first initialize a database storage area on disk.

.. code-block:: console

    $ initdb -D mylocal_db

This *database cluster* may contain a collection of databases (one per profile) that is managed by a single running server process.
We start this process with:

.. code-block:: console

    $ pg_ctl -D mylocal_db -l logfile start

.. admonition:: Further Reading
    :class: seealso title-icon-read-more

    - `Creating a Database Cluster <https://www.postgresql.org/docs/12/creating-cluster.html>`__.
    - `Starting the Database Server <https://www.postgresql.org/docs/12/server-start.html>`__.

> To set up an AiiDA configuration profile and related data storage, we can then use the `quicksetup` command.

Reviewer: ```suggestion
Next, we set up an AiiDA configuration profile and related data storage, with the `quicksetup` command.
```

Getting Started
****************

A working AiiDA installation consists of three core components, plus any external codes you wish to run:

* ``aiida-core``: The main python package and associated CLI ``verdi``.
* |PostgreSQL|: A service which manages the database where we store generated data.
* |RabbitMQ|: A service which manages communication with the processes that we run.

AiiDA provides a number of routes to setting up this environment, depending on your use case.
These are outlined below, followed by a recommended "quick-install" route on your local computer.

.. panels::
   :body: bg-light
   :footer: bg-light border-0

   :fa:`desktop,mr-1` **Direct (Bare Metal)**

   *Install software directly into your local root directory.*

   The required software can be installed using most package managers, including: apt, Homebrew, MacPorts, Gentoo and Windows Subsystem for Linux.

   +++

   :link-badge:`install/software,Software install,ref,badge-primary text-white`
   :link-badge:`install/aiida-core,aiida-core install,ref,badge-primary text-white`

   ---------------

>    :fa:`folder,mr-1` **Virtual Folder**

Author: Yeh agree that the common terminology around this is Virtual Environment. However, my only thinking here was that this terminology is a bit misleading, when also considering containers and VMs. They are actually virtual, in the sense that all processes are isolated) whereas really these "virtual" environments are just "placing everything in the same folder".

At this point you now have a working AiiDA environment, from which you can add and retrieve data.

.. admonition:: Tab Completion
    :class: tip title-icon-lightbulb

    Enable tab completion of ``verdi`` commands in the terminal with:

    .. code-block:: console

        $ eval "$(_VERDI_COMPLETE=source verdi)"

    :link-badge:`how-to:installation:configure:tab-completion,Read More,ref,badge-primary text-white`

In order to run computations, some additional steps are required to start the services that manage these background processes.

> The |RabbitMQ| service is started, to manage communication between processes and remember process states, even when you shut down your computer:

Reviewer: ```suggestion
The |RabbitMQ| service is used, to manage communication between processes and retain process states, even after restarting your computer:
```

.. code-block:: console

    $ rabbitmq-server -detached

> We can then start a one or more "daemon" processes, which handle starting and monitoring all submitted computations.

Reviewer: ```suggestion
We then start one or more "daemon" processes, which handle the execution and monitoring of all submitted computations.
```

Awesome! You now have a fully operational installation from which to take the next steps!

After finishing with your aiida session, you may wish to power down the services:

.. code-block:: console

    $ verdi daemon stop
    $ pg_ctl stop

If computations are still running, then they will be picked up next time the services are started.

.. admonition:: Having problems?
    :class: attention title-icon-troubleshoot

    :ref:`See the troubleshooting section <intro/troubleshooting>`

.. admonition:: In-depth instructions
    :class: tip title-icon-read-more

    For more ways to install AiiDA :ref:`Check the advanced installation section <intro/installation>`.

    For more detailed instructions on configuring AiiDA, :ref:`see the configuration instructions <configure_aiida>`

Next Steps
==========

.. div:: dropdown-group

    .. dropdown:: Run pure Python lightweight computations
        :container:

>         blah blah blah

Author: See: https://65-77234579-gh.circle-artifacts.com/0/html/intro/get_started.html#next-steps

Awesome! You now have a fully operational installation from which to take the next steps!

After finishing with your aiida session, you may wish to power down the services:

.. code-block:: console

    $ verdi daemon stop
    $ pg_ctl stop

> If computations are still running, then they will be picked up next time the services are started.

Reviewer: ```suggestion
Any computations that are still running at this point, will be picked up next time the services are started.
```

.. admonition:: Having problems?
    :class: attention title-icon-troubleshoot

    :ref:`See the troubleshooting section <intro:troubleshooting>`.

.. admonition:: In-depth instructions
    :class: seealso title-icon-read-more

>     For more ways to install AiiDA, :ref:`check the advanced installation section <intro:install>`.

Reviewer: ```suggestion
    For more ways to install AiiDA, :ref:`check the detailed installation section <intro:install>`.
```

    For more detailed instructions on configuring AiiDA, :ref:`see the configuration how-to <how-to:installation:configure>`.

Next Steps
==========

> Try out the :ref:`Basic Tutorial <tutorial:basic>`, or check if your use case is one of the following:

Reviewer: ```suggestion
If you are new to AiiDA, go through the :ref:`Basic Tutorial <tutorial:basic>`.
We also compiled some useful how-to guides that are especially relevant for the following use cases:
```

        Exploring your data
            Once you have run multiple computations, the :ref:`find and query data how-to <how-to:data:find>` can show you how to efficiently explore your data. The data lineage can also be visualised as a :ref:`provenance graph <VisualizingGraphs>`.

        Sharing your data
            You can export all or part of your data to file with the :ref:`export/import functionality<how-to:data:share>` or you can even serve your data over HTTP(S) using the :ref:`AiiDA REST API <how-to:data:serve>`.

    blah blah blah
        Sharing your workflows
            Once you have a working computation workflow, you may also wish to :ref:`package it into a python module <how-to:plugins>` for others to use.

    .. container:: link-box
    .. dropdown:: Run compute-intensive codes
        :container:

        links to tutorials
        Tuning performance
            To optimise the performance of AiiDA for running many concurrent computations see the :ref:`tuning performance how-to <how-to:installation:performance>`.

> .. accordion:: Run compute-intensive codes

Author: fixed in 5264eb829b1050b5a9073a4f2a382d5d5c638a33
